samples: nrf_rpc: protocols_serialization: documentation restructuring (#16597)
nrfconnect/sdk-nrf: samples/nrf_rpc/protocols_serialization/server/README.rst

.. _nrf_rpc_protocols_serialization_server:

Protocols serialization: server
###############################

.. contents::
   :local:
   :depth: 2

The Protocols serialization server sample runs full Bluetooth® LE and OpenThread stacks, and it exposes selected functions from these stacks over UART.

Requirements
************

The sample supports the following development kits for testing the network status:

.. table-from-sample-yaml::

To test the sample, you also need another device running the :ref:`Protocols serialization client <nrf_rpc_protocols_serialization_client>` sample.

For testing the Bluetooth LE API serialization, you also need to have the `nRF Connect for Mobile`_ app installed on your smartphone or tablet.

Overview
********

The Protocols serialization server sample implements the server part of the serialization of OpenThread and Bluetooth LE API calls between two devices that communicate with each other using the UART interface.
The sample uses the :ref:`nrfxlib:nrf_rpc` and CBOR encoding to serialize function calls.

Configuration
*************

> Snippets

Reviewer: ```suggestion
|config|

Snippets
```

========

> * set the :makevar:`SNIPPET` option to ``ble;coex;debug;log_rpc;openthread``

Reviewer: Same comment as for the client sample.

> .. include:: /includes/sample_snippets.txt

Reviewer: ```suggestion
.. |snippet| replace:: :makevar:`server_SNIPPET`

.. include:: /includes/sample_snippets.txt
```

The following snippets are available:

> * ``ble`` - Enables the server part of the :ref:`Bluetooth LE RPC <ble_rpc>`. Enables the :ref:`Bluetooth LE Nordic UART Service <nus_service_readme>`.

Reviewer: ```suggestion
* ``ble`` - Enables the server part of the :ref:`Bluetooth LE RPC <ble_rpc>`.
  Also enables the :ref:`Bluetooth LE Nordic UART Service <nus_service_readme>`.
```

You can modify the list of enabled features, which by default includes Bluetooth LE support and debug logs.

.. _protocols_serialization_server_sample_testing:

Testing
*******

After building the Protocols serialization server sample and programming it to your development kit, connect it to a second device running the :ref:`Protocol serialization client <nrf_rpc_protocols_serialization_client>` sample to test either the Bluetooth LE or OpenThread functionality.

.. _protocols_serialization_server_app_connection:

Connecting the client and server samples
========================================

The client and server devices are connected using two UART peripherals.
In the protocols serialization samples, one peripheral is used for shell and logging purposes, similarly to other applications and samples, while the other peripheral is used for sending OpenThread and Bluetooth LE remote procedure calls (RPCs).

.. tabs::

    .. group-tab:: nRF52840 SoC

        By default, the nRF52840 system-on-chip uses the ``uart0`` peripheral for shell and logging purposes, and the ``uart1`` peripheral for sending OpenThread and Bluetooth remove procedure calls (RPCs).

>         The `uart1` peripheral is configured to use the following pins:

Reviewer: ```suggestion
        The ``uart1`` peripheral is configured to use the following pins:
```

        - **P1.1**: RX
        - **P1.2**: TX
        - **P1.3**: RTS (hardware flow control)
        - **P1.4**: CTS (hardware flow control)

        Therefore, in order to enable the communication between the client and the server devices, connect the two nRF 52840 DKs using jumper wires according to the following diagram:

        .. figure:: /images/ps_nrf52_connections.png
            :alt: nRF52 connections

    .. group-tab:: nRF54l15 SoC

        By default, the nRF54L15 system-on-chip uses the ``uart20`` peripheral for shell and logging purposes, and the ``uart21`` peripheral for sending OpenThread and Bluetooth remove procedure calls (RPCs).

>         The `uart21` peripheral is configured to use the following pins:

Reviewer: ```suggestion
        The ``uart21`` peripheral is configured to use the following pins:
```

* ``openthread`` - Enables the server part of the OpenThread RPC.

Building and running
********************

.. |sample path| replace:: :file:`samples/nrf_rpc/protocols_serialization/server`

.. include:: /includes/build_and_run.txt

You can modify the list of enabled features, which by default includes Bluetooth LE support and debug logs.

.. _protocols_serialization_server_sample_testing:

Testing
*******

After building the Protocols serialization server sample and programming it to your development kit, connect it to a second device running the :ref:`Protocol serialization client <nrf_rpc_protocols_serialization_client>` sample to test either the Bluetooth LE or OpenThread functionality.

.. _protocols_serialization_server_app_connection:

Connecting the client and server samples
========================================

The client and server devices are connected using two UART peripherals.
In the protocols serialization samples, one peripheral is used for shell and logging purposes, similarly to other applications and samples, while the other peripheral is used for sending OpenThread and Bluetooth LE remote procedure calls (RPCs).

.. tabs::

>     .. group-tab:: nRF52840 SoC
> 
>         By default, the nRF52840 system-on-chip uses the ``uart0`` peripheral for shell and logging purposes, and the ``uart1`` peripheral for sending OpenThread and Bluetooth remove procedure calls (RPCs).

Reviewer: ```suggestion
    .. group-tab:: nRF52840 DK

        By default, the nRF52840 DK uses the ``uart0`` peripheral for shell and logging purposes, and the ``uart1`` peripheral for sending OpenThread and Bluetooth remove procedure calls (RPCs).
```
We're showing pictures of the DK, so might as well call it the DK instead of SoC.

        The ``uart1`` peripheral is configured to use the following pins:

>         - **P1.1**: RX
>         - **P1.2**: TX
>         - **P1.3**: RTS (hardware flow control)
>         - **P1.4**: CTS (hardware flow control)
> 
>         Therefore, in order to enable the communication between the client and the server devices, connect the two nRF 52840 DKs using jumper wires according to the following diagram:

Reviewer: ```suggestion
        .. list-table::
           :header-rows: 1
           
           * - Server
             - Client
             - Function on server
           * - **P1.1**
             - **P1.2**
             - RX
           * - **P1.2**
             - **P1.1**
             - TX
           * - **P1.3**
             - **P1.4**
             - RTS (hardware flow control)
           * - **P1.4**
             - **P1.3**
             - CTS (hardware flow control)
           * - **GND**
             - **GND**
             - Ground

        To enable the communication between the client and the server devices, connect the pins on the two nRF52840 DKs using jumper wires.
        The following illustration demonstrates the pin connections:
```

        .. figure:: /images/ps_nrf52_connections.png

>             :alt: nRF52 connections

Reviewer: ```suggestion
            :alt: nRF52840 DK server and client pin connections
            
            nRF52840 DK server and client pin connections
```

>     .. group-tab:: nRF54l15 SoC
> 
>         By default, the nRF54L15 system-on-chip uses the ``uart20`` peripheral for shell and logging purposes, and the ``uart21`` peripheral for sending OpenThread and Bluetooth remove procedure calls (RPCs).

Reviewer: ```suggestion
    .. group-tab:: nRF54l15 DK

        By default, the nRF54L15 DK uses the ``uart20`` peripheral for shell and logging purposes, and the ``uart21`` peripheral for sending OpenThread and Bluetooth remove procedure calls (RPCs).
```

        The ``uart21`` peripheral is configured to use the following pins:

>         - **P1.10**: RX
>         - **P1.11**: TX
>         - **P1.12**: RTS (hardware flow control)
>         - **P1.13**: CTS (hardware flow control)

Reviewer: ```suggestion
        .. list-table::
           :header-rows: 1
           
           * - Server
             - Client
             - Function on server
           * - **P1.10**
             - **P1.11**
             - RX
           * - **P1.11**
             - **P1.10**
             - TX
           * - **P1.12**
             - **P1.13**
             - RTS (hardware flow control)
           * - **P1.13**
             - **P1.12**
             - CTS (hardware flow control)
           * - **GND**
             - **GND**
             - Ground

        To enable the communication between the client and the server devices, connect the pins on the two nRF54L15 DKs using jumper wires.
        The following illustration demonstrates the pin connections:
```

        .. figure:: /images/ps_nrf54l_connections.png

>             :alt: nRF54l connections

Reviewer: ```suggestion
            :alt: nRF54L15 DK server and client pin connections
            
            nRF54L15 DK server and client pin connections
```

Testing
*******

For testing the server sample, use the protocol serialization client sample :ref:`nrf_rpc_protocols_serialization_client`.

.. _protocols_serialization_server_app_connection:

Connecting client and server
============================

The client and server devices are connected using the UART peripherals.

.. tabs::

    .. group-tab:: nRF52840

        The nRF52840 system-on-chip includes two UART peripherals: `uart0` and `uart1`. In the protocols serialization sample, the `uart0` peripheral is used for shell and logging purposes, similarly to other applications and samples, while the `uart1` peripheral is used for sending OpenThread and Bluetooth LE remote procedure calls (RPCs).

        The `uart1` peripheral is configured to use the following pins:

        - **P1.1**: RX
        - **P1.2**: TX
        - **P1.3**: RTS (hardware flow control)
        - **P1.4**: CTS (hardware flow control)

        Therefore, in order to enable the communication between the client and the server devices, connect the two nRF 52840 DKs using jumper wires according to the following diagram:

        .. figure:: /images/ps_nrf52_connections.png

>             :alt: nRF52 connections

Author: Kind of disagree. Picture provides better visibility how to connect. As a customer, I do not need to count the number of pins and read what is written on the board. It is sufficient just connect to each other shown pins. 
Since I came to the project after it was drawn, and I was the first customer of this documentation, I would say that the pictures are much convenient.

>             :alt: nRF52 connections

Reviewer: I would be worried that a customer that looks at the illustration and does not do the counting and checking names might end up connecting the pins wrong, and then blame Nordic for inaccurate instructions.

But I think we can compromise here. If we replace the bullet lists of pins with the table and mention that **GND** need to be connected, and then introduce the illustrations with "The following figure illustrates how these pins are connected:", it makes it clear that the illustration is a secondary source.

Even then, I'd then like to have the illustrations cleaned up so the connecting lines are clean and use colors from Nordic color system. I'll try to produce versions either today or tomorrow.

    .. group-tab:: nRF54l15

        The nRF54l15 system-on-chip uses by default `uart21` for RPC and `uart20` is used for shell.

        The `uart21` peripheral is configured to use the following pins:

        - **P1.10**: RX
        - **P1.11**: TX
        - **P1.12**: RTS (hardware flow control)
        - **P1.13**: CTS (hardware flow control)

        .. figure:: /images/ps_nrf54l_connections.png

>             :alt: nRF54l connections

Author: https://github.com/nrfconnect/sdk-nrf/pull/16597#discussion_r1711464809

> .. _protocols_serialization_server_sample_testing:

Author: Why? Seem testing tag always goes before Testing chapter. Please clarify why this sample requires different approach.

#. Connect to the UART console of the client device using a terminal emulator of your choice.

#. Connect to the UART console of the server device using a terminal emulator of your choice.

#. Reboot **both** devices at the same time using the **IF BOOT/RESET** button on the DK.

#. Wait a few seconds until similar message on both UART consoles:

   .. code-block:: console

      [00:00:00.842,862] <inf> nrf_rpc_host: RPC client ready

   This indicates that the communication between the devices has been initialized properly.

#. Run the following command on the client's console to start Bluetooth LE advertising:

   .. code-block:: console

      uart:~$ bt init
      uart:~$ bt advertise on

#. On mobile phone, open `nRF Connect` app and scan for `Nordic_UART_Service` as in the image below:

   .. figure:: /images/ps_nrf_scan.png
      :alt: BLE scanning

#. Tap **CONNECT** button to connect to the client device.
   Observe the following message on the client's console:

   .. code-block:: console

      LE conn param updated: int 0x0027 lat 0 to 42

#.  Tap the arrow next to the RX characteristic to send data:

   .. figure:: /images/ps_nrf_send.png

>       :alt: BLE send data

Author: I agree with you on this. When I was running the sample I had iPhone not Android and even didn't look at the picture to do what is required. This is documentation for server sample, not for mobile application. The mobile application should have its own documentation or just intuitive interface.
I'll remove picture and rephrase this paragraph.

   Observe the following message on the client's console:

   .. code-block:: console

      bt_nus: on_receive: Received data, handle 0, conn 0x200023c4

Testing OpenThread
==================

Complete the following steps to test OpenThread API serialization:

#. Connect to the UART console of the client device using a terminal emulator of your choice.

#. Connect to the UART console of the server device using a terminal emulator of your choice.

#. Reboot **both** devices at the same time using the **IF BOOT/RESET** button on the DK.

> #. Wait a few seconds until the following message on both UART consoles:

Reviewer: ```suggestion
#. Wait a few seconds until you see a message similar to the following on both terminal emulators:
```

   .. code-block:: console

      uart:~$ [00:00:03.392,517] <dbg> NRF_RPC: nrf_rpc_init: Done initializing nRF RPC module

   This indicates that the communication between the devices has been initialized properly.

> #. Run the following command on the client's console to bring the OpenThread interface on the server device up:

Reviewer: ```suggestion
#. Run the following command on the client's terminal emulator to bring up the OpenThread interface on the server device:
```

   .. code-block:: console

      uart:~$ ot ifconfig up
      Done
      [00:02:28.980,041] <dbg> NRF_RPC: cmd_ctx_alloc: Command context 0 allocated
      [00:02:28.980,102] <dbg> NRF_RPC: nrf_rpc_cmd_common: Sending command 0x00 from group 0x01
      [00:02:28.980,133] <dbg> nrf_rpc_uart: send: Sending frame
                                             80 00 ff 01 01 f6
      ...

> #. Bring the corresponding Zephyr network interface on the client device:

Reviewer: ```suggestion
#. Run the following command on the client's terminal emulator to bring up the corresponding Zephyr network interface on the client device:
```

      =========================================
      MTU       : 1280
      Flags     : NO_AUTO_START,IPv6,NO_ND,NO_MLD
      Device    : ot_rpc (0x2b748)
      IPv6 unicast addresses (max 5):
            fe80::6c26:956a:813:1e33 autoconf preferred infinite
            fdde:ad00:beef:0:e503:abfd:1c8d:2664 autoconf preferred infinite meshlocal
            fdde:ad00:beef::ff:fe00:fc00 autoconf preferred infinite meshlocal

   This happens because the client registers a notification callback for OpenThread state changes at the server device and it continuously refreshes the client's IPv6 address list when that changes on the server side.

#. Retrieve the operational dataset of the Thread network:

    .. code-block:: console

      uart:~$ ot dataset active -x
      0e080000000000000000000300000b35060004001fffe00208dead00beef00cafe0708fddead00beef00000510f7893f15a55d8adeacad288c38bf32cc030a4f70656e546872656164010240120410d48d777a474f80e61aa5680de764bd6d0c0402a0f7f8
      Done

#. To send a UDP packet to a peer device and UDP port 5555, run the following command:

   .. code-block:: console

      uart:~$ net udp send fe80:0:0:0:6c26:956a:813:1e34 5555 AAAA
      Message sent

> #. To open a UDP port 5555 and listen for incoming UDP datagrams from peer devices, run the following command:

Reviewer: ```suggestion
#. To open the UDP port 5555 and listen for incoming UDP datagrams from peer devices, run the following command:
```

   .. code-block:: console

      uart:~$ net udp bind fe80:0:0:0:6c26:956a:813:1e33 5555

> Please note that, all IPv6 addresses are just examples and could be replaced with the actual IPv6 addresses of the peer devices.

Reviewer: ```suggestion
Note that all IPv6 addresses shown here are just examples and will be replaced with the actual IPv6 addresses of the peer devices.
```
